### DISCUSSION TITLE
What is the encoding order of Map and Set elements?

### DISCUSSION THREAD

**Discussion by streetycat** [buckyos/CYFS#253] [Q&A]
I am reviewing the encoding format for types. But I find that our encoding process only explicitly sorts the HashMap. Other types(as HashSet/BTreeMap/BTreeSet) depends on the iter in `Rust`. I have read the documents for those types, The iters for BTreeMap and BTreeSet return elements in ascending order. but I have not found how it is sorted in HashSet.

I am reading the code in [raw_types.rs](https://github.com/buckyos/CYFS/blob/main/src/component/cyfs-base/src/codec/raw/raw_types.rs).

I think we have to explicitly define the encoding order of the fields/elements of the composite type, otherwise different results may be encoded.

**Comment by lurenpluto** [comment]
# The stability of the container encoding problem
To be precise, object codec to container encoding, are converted to an array after encoding, so after conversion to an array involves the issue of element order

BTreeMap/BTreeSet itself is traversed in order, sorting depends on the corresponding Ord trait design, custom Object can design their own Ord trait impls, as long as the algorithm to ensure compliance with the container  [Ord](https://doc.rust-lang.org/std/cmp/trait.Ord.html)

The elements inside the HashMap/HashSet do not need to support Ord trait, but for the need of stable coding, the default implementation of RawCodec restricts the need to support Ord this trait:
https://github.com/buckyos/CYFS/blob/c7a777dd7dc06ac9f93a84a10517d864fb0a758d/src/component/cyfs-base/src/codec/raw/raw_types.rs#L784-L792

So the internal sorting based on the elements Ord, now it seems a little strange, if the user's HashMap element type does not support Ord trait, then it will not be able to use the default sort and RawCodec impls, and the user may not need a stable sort itself,  and there is a performance waste here itself

**HashSet this one should be missed, according to the principle of consistency, should be consistent with the HashMap**

# Design principles
I think this we should have the following design principles

1. If you need stable sorting and rely on RawCodec's default codec implementation, then please use BTreeMap/BTreeSet
2. If you don't need stable sorting, you can also choose HashMap/HashSet instead
3. If you want to use the implementation of HashMap/HashSet and also want to achieve stable sorting, then please implement the codec of the stable version of the corresponding data structure by yourself.

# Introduce the label
Therefore, there should be a label "stable encoding or not" to identify the default data type supported by cyfs-base, so that you can choose the stable or non-stable version of the data structure according to your needs when designing the data structure for object selection.

**Comment by streetycat** [comment]
I think we should only define a definite arrangement order, we just need to ensure that the encoded content of the same content is exactly the same. For the user, each element is stored in the container after decoding, and how the elements are arranged in it is not important.

**Reply by lurenpluto** [reply]
> I think we should only define a definite arrangement order, we just need to ensure that the encoded content of the same content is exactly the same. For the user, each element is stored in the container after decoding, and how the elements are arranged in it is not important.

I think there are two points to consider

1. The user-defined encoding of the elements, does not necessarily support sorting, that is,  elements does not support `Ord` trait, for this case there is no way to follow our "codec specification" as objectfields
2. Object's body encoding, can not need to be stable (except for the need to sign, otherwise it may lead to failure of verification), such as just in the local NOC storage

However, simply specifying the order of encoding can avoid some potential problems
